File: BaseSDK/AArch64/Samples/Base/DS86/DeviceParamSetGet/DeviceParamSetGet.cpp
#include <thread>
#include <iostream>
#include "Scepter_api.h"

using namespace std;

int main()
{
	cout << "---DeviceParamSetGet---" << endl;

	uint32_t deviceCount;
	ScDeviceInfo* pDeviceListInfo = NULL;
	ScDeviceHandle deviceHandle = 0;
	ScStatus status = SC_OTHERS;

	status = scInitialize();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scInitialize] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scInitialize] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scGetDeviceCount(&deviceCount, 3000);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceCount] success, ScStatus(" << status << "). The device count is " << deviceCount << endl;
	}
	else
	{
		cout << "[scGetDeviceCount] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}
	if (0 == deviceCount)
	{
		cout << "[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples." << endl;
		return -1;
	}

	pDeviceListInfo = new ScDeviceInfo[deviceCount];
	status = scGetDeviceInfoList(deviceCount, pDeviceListInfo);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetDeviceInfoList] success, ScStatus(" << status << ").";
		if (SC_CONNECTABLE != pDeviceListInfo[0].status)
		{
			cout << " The first device [status]: " << pDeviceListInfo[0].status << " does not support connection." << endl;
			delete[] pDeviceListInfo;
			pDeviceListInfo = NULL;
			return -1;
		}
	}
	else
	{
		cout << "[scGetDeviceInfoList] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return -1;
	}

	cout << " The first deviceInfo, <serialNumber>: " << pDeviceListInfo[0].serialNumber
		<< ", <ip>: " << pDeviceListInfo[0].ip << ", <status>: " << pDeviceListInfo[0].status << endl;

	status = scOpenDeviceBySN(pDeviceListInfo[0].serialNumber, &deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scOpenDeviceBySN] success, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
	}
	else
	{
		cout << "[scOpenDeviceBySN] fail, ScStatus(" << status << ")." << endl;
		delete[] pDeviceListInfo;
		pDeviceListInfo = NULL;
		return -1;
	}

	ScSensorIntrinsicParameters cameraParameters;
	status = scGetSensorIntrinsicParameters(deviceHandle, SC_TOF_SENSOR, &cameraParameters);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetSensorIntrinsicParameters] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scGetSensorIntrinsicParameters] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	cout << endl;
	cout << "Depth camera intinsic: " << endl;
	cout << "Fx: " << cameraParameters.fx << endl;
	cout << "Cx: " << cameraParameters.cx << endl;
	cout << "Fy: " << cameraParameters.fy << endl;
	cout << "Cy: " << cameraParameters.cy << endl;

	cout << "Depth distortion coefficient: " << endl;
	cout << "K1: " << cameraParameters.k1 << endl;
	cout << "K2: " << cameraParameters.k2 << endl;
	cout << "P1: " << cameraParameters.p1 << endl;
	cout << "P2: " << cameraParameters.p2 << endl;
	cout << "K3: " << cameraParameters.k3 << endl;
	cout << "K4: " << cameraParameters.k4 << endl;
	cout << "K5: " << cameraParameters.k5 << endl;
	cout << "K6: " << cameraParameters.k6 << endl;
	cout << endl;

	uint8_t gmmgain = 0;
	status = scGetIRGMMGain(deviceHandle, &gmmgain);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scGetIRGMMGain] success, ScStatus(" << status << "). The device IRGMM gain is " << (uint32_t)gmmgain << endl;
	}
	else
	{
		cout << "[scGetIRGMMGain] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	gmmgain = 50;
	status = scSetIRGMMGain(deviceHandle, gmmgain);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scSetIRGMMGain] success, ScStatus(" << status << "). Set the device IRGMM gain to " << (uint32_t)gmmgain << endl;
	}
	else
	{
		cout << "[scSetIRGMMGain] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scCloseDevice(&deviceHandle);
	if (status == ScStatus::SC_OK)
	{
		cout << "[scCloseDevice] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scCloseDevice] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	status = scShutdown();
	if (status == ScStatus::SC_OK)
	{
		cout << "[scShutdown] success, ScStatus(" << status << ")." << endl;
	}
	else
	{
		cout << "[scShutdown] fail, ScStatus(" << status << ")." << endl;
		return -1;
	}

	cout << "---End---" << endl;
	return 0;
}

File: MultilanguageSDK/Python/Samples/DS86/DeviceParamSetGet/DeviceParamSetGet.py
from pickle import FALSE, TRUE
import sys
currentPath =  sys.path[0]
pos = currentPath.find('Samples')
libpath = currentPath[:pos]
sys.path.append(libpath) #absolutely path

from API.ScepterDS_api import *
import time
print("---DeviceParamSetGet---")
camera = ScepterTofCam()

camera_count = camera.scGetDeviceCount(3000)
print("[scGetDeviceCount] success, ScStatus(0), The device count is {}".format(camera_count))
if camera_count <= 0:
    print("[scGetDeviceCount] scans for 3000ms and then returns the device count is 0. Make sure the device is on the network before running the samples.")
    exit(1)

device_info=ScDeviceInfo()

if camera_count > 0:
    ret,device_infolist=camera.scGetDeviceInfoList(camera_count)
    if ret==0:
        device_info = device_infolist[0]
        print("[scGetDeviceInfoList] success, ScStatus({}). The first deviceInfo, <serialNumber>: {} , <ip>: {} , <status>: {}".format(ret, str(device_info.serialNumber.decode()), str(device_info.ip.decode()), str(device_info.status)))
        if  ScConnectStatus.SC_CONNECTABLE.value != device_info.status:
            print(" The first device [status]: {} does not support connection.".format(str(device_info.status)))
            exit(1)
    else:
        print("[scGetDeviceInfoList] fail, ScStatus({})".format(ret))
        exit(1) 
else: 
    print("there are no camera found")
    exit(1)

ret = camera.scOpenDeviceBySN(device_info.serialNumber)
if  ret == 0:
    print('[scOpenDeviceBySN] success ScStatus({}).'.format(str(ret)))
else:
    print('[scOpenDeviceBySN] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret, params = camera.scGetSensorIntrinsicParameters(ScSensorType.SC_TOF_SENSOR)
if  ret == 0:
    print('[scGetSensorIntrinsicParameters] success ScStatus({}).'.format(str(ret)))
    print("Depth camera intinsic: \nFx: {}\nCx: {}\nFy: {}\nCy: {}".format(params.fx, params.cx, params.fy, params.cy))
    print("Depth distortion coefficient: \nK1: {}\nK2: {}\nP1: {}\nP2: {}\nK3: {}\nK4: {}\nK5: {}\nK6: {}".format(params.k1,params.k2,params.p1,params.p2,params.k3,params.k4,params.k5,params.k6))
else:
    print('[scGetSensorIntrinsicParameters] fail ScStatus({}).'.format(str(ret)))

ret, gmmgain = camera.scGetIRGMMGain()
if  ret == 0:
    print('[scGetIRGMMGain] success ScStatus({}). The device IRGMM gain is {}.'.format(str(ret), gmmgain))
else:
    print('[scGetIRGMMGain] fail ScStatus({}).'.format(str(ret)))
    exit(1)

gmmgain = 50
ret = camera.scSetIRGMMGain(gmmgain)
if  ret == 0:
    print('[scSetIRGMMGain] success ScStatus({}). Set the device IRGMM gain to {}.'.format(str(ret), gmmgain))
else:
    print('[scSetIRGMMGain] fail ScStatus({}).'.format(str(ret)))
    exit(1)

ret = camera.scCloseDevice()
if  ret == 0:
    print('[scCloseDevice] success ScStatus({}).'.format(str(ret)))
else:
    print('[scCloseDevice] fail ScStatus({}).'.format(str(ret)))
    exit(1)

exit(0)
